### PR TITLE
[Enhancement] Support multi count distinct with different multi cloumns

### LINKED ISSUE #7495: Support multi count distinct with different multi cloumns
## Enhancement
StarRocks do not support multi count distinct with different multi cloumns
such like:
```
mysql> select count(distinct c_address,c_nationkey),count(distinct c_address),c_nationkey from customer group by c_nationkey;
ERROR 1064 (HY000): The query contains multi count distinct or sum distinct, each can't have multi columns.
```

### PR DESCRIPTION
**Pull request by Youngwb**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #7495 

## Problem Summary(Required) ：

Use CTE to support sql like : 
count(distinct a,b), count(distint a)

plan like ：
```
Mark count(distinct v1) as cd(v1), avg(distinct v1) as ad(v1), sum(distinct v1) as sd(v1).
 * Before:
 *          Agg[cd(v1), cd(v2), sum(v4), max(v5)] group by v3
 *                      |
 *                  Child Plan
 *
 * After:
 *                     CTEAnchor
 *                   /            \
 *           CTEProduce          inner join (v3 = v3)
 *             /                 /                  \
 *      Child Plan        inner join(v3 = v3)       Agg[sum(v4), max(v5)]
 *                        /          \               group by v3
 *                       /            \                   \
 *                 Agg[cd(v1)]     Agg[cd(v2)]        CTEConsume
 *                 group by v3     group by v3
 *                     /               \
 *                CTEConsume        CTEConsume
 *
```

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 97 / 97 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctWithoutGroupByRule.java|96|96|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 48 / 48 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctWithoutGroupByRule.java|47|47|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|1|1|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctWithoutGroupByRule.java [hunk 1]
```diff
@@ -78,6 +82,24 @@
...
+ *                        /          \               group by v3
+ *                       /            \                   \
+ *                 Agg[cd(v1)]     Agg[cd(v2)]        CTEConsume
+ *                 group by v3     group by v3
+ *                     /               \
+ *                CTEConsume        CTEConsume
+ *
  *
  */
 public class RewriteMultiDistinctWithoutGroupByRule extends TransformationRule {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctWithoutGroupByRule.java:105]: update name

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 48 / 48 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|47|47|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|1|1|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 48 / 48 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctByCTERule.java|47|47|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RewriteMultiDistinctRule.java|1|1|100.00%|

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by Youngwb** [comment]
@Mergifyio backport branch-2.3

**Comment by ShepardZhao** [comment]
this is awesome

**Pull request merged**